### PR TITLE
Push new build name into environment

### PR DESCRIPTION
**Pull request by Le0Michine**
Added new feature.
New build name will be stored into environment variable so it is possible to see it from another build step using ${BUILD_DISPLAY_NAME}

It is also fix for bug JENKINS-33687 https://issues.jenkins-ci.org/browse/JENKINS-33687

### CODE REVIEW

File: src/main/java/org/jenkinsci/plugins/EnvironmentHelper.java [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+import java.util.logging.Logger;
+
+/**
+ * Created by Leo on 4/20/2016.
+ * Helper to work with environment variables.
+ */
+public class EnvironmentHelper {
+    private static final Logger LOGGER = Logger.getLogger(EnvironmentHelper.class.getName());
+
+	public static final String BuildDisplayNameVar = "NEW_BUILD_NAME";
```
Comment (oleg-nenashev) [inline src/main/java/org/jenkinsci/plugins/EnvironmentHelper.java]: NIT: formatting

File: src/main/java/org/jenkinsci/plugins/EnvironmentHelper.java [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+	
+    public static void SetEnvironmentVariable(@CheckForNull String key, @CheckForNull String value, PrintStream logger) {
```
Comment (bsideup) [inline src/main/java/org/jenkinsci/plugins/EnvironmentHelper.java]: in Java, methods names are in camelCase. Other methods in build-name-setter-plugin are also in camelCase

Comment (bsideup) [inline src/main/java/org/jenkinsci/plugins/EnvironmentHelper.java]: you can add @CheckForNull to `logger` as well since usage is not null-safe

Comment (bsideup) [inline src/main/java/org/jenkinsci/plugins/EnvironmentHelper.java]: also, SetEnvironmentVariable accepts `logger`, but in GetEnvironmentVariable `LOGGER` is being used. Makes sense to use `GetEnvironmentVariable` in SetEnvironmentVariable as well. What do you think?

Comment (Le0Michine) [inline src/main/java/org/jenkinsci/plugins/EnvironmentHelper.java]: thanks, .net habit :)

File: src/main/java/org/jenkinsci/plugins/buildnameupdater/BuildNameUpdater.java [hunk 3]
```diff
@@ -161,7 +162,6 @@ public DescriptorImpl getDescriptor() {
     private static class MyFileCallable implements FilePath.FileCallable<String> {
         private static final long serialVersionUID = 1L;
 
-        @Override
```
Comment (bsideup) [inline src/main/java/org/jenkinsci/plugins/buildnameupdater/BuildNameUpdater.java]: why did you remove `@Override` from here?

**Comment by oleg-nenashev** [comment]
I do not understand the logic and purpose of this PR.
You save the variable into global node properties, so the value will be available from ANY build. And this behavior will be a subject for all kinds of concurrency issues.

The only thing you need is to add [EnvironmentContributingAction](http://javadoc.jenkins-ci.org/hudson/model/EnvironmentContributingAction.html) to the build from your build step.

File: src/main/java/org/jenkinsci/plugins/buildnameupdater/BuildNameUpdater.java (hunk 3 above)
Comment (Le0Michine) [inline src/main/java/org/jenkinsci/plugins/buildnameupdater/BuildNameUpdater.java]: because it is interface implementation and was marked as error

**Comment by Le0Michine** [comment]
PR isn't ready to be merged, I'll rework it soon

File: src/main/java/org/jenkinsci/plugins/buildnameupdater/BuildNameUpdater.java (hunk 3 above)
Comment (bsideup) [inline src/main/java/org/jenkinsci/plugins/buildnameupdater/BuildNameUpdater.java]: Starting with Java 6 it's allowed and recommended to put `@Override` on implementations of interface methods, see this discussion for explanation:

http://stackoverflow.com/questions/212614/should-we-override-an-interfaces-method-implementation

Comment (Le0Michine) [inline src/main/java/org/jenkinsci/plugins/buildnameupdater/BuildNameUpdater.java]: OK, I got it, thanks!

**Comment by Le0Michine** [comment]
@oleg-nenashev @bsideup ping :)

File: src/main/java/org/jenkinsci/plugins/EnvironmentVarSetter.java [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+
+        log.println("Create var " + key + "=" + value);
+        if (envVars.containsKey(key)) {
+            log.println("Current value '" + envVars.get(key) + "'");
+        }
+        envVars.put(key, value);
+    }
+
+    public String getVar(String key) {
+        log.println("Get var '" + key + "'");
```
Comment (bsideup) [inline src/main/java/org/jenkinsci/plugins/EnvironmentVarSetter.java]: I think you're overusing logging here. Since it has no "log level" semantics, both debug and info messages will appear. Doesn't seems valuable to me

File: src/test/java/org/jenkinsci/plugins/buildnamesetter/BuildNameSetterTest.java [hunk 5]
```diff
@@ -78,6 +79,8 @@ public void shouldUse_default_config_values_if_null() throws InterruptedExceptio
 	private void asssertDisplayName(FreeStyleBuild build, String expectedName) {
 		assertEquals(Result.SUCCESS, build.getResult());
 		assertEquals(expectedName, build.getDisplayName());
+		EnvironmentVarSetter action = build.getAction(EnvironmentVarSetter.class);
+		assertEquals(expectedName, action.getVar(EnvironmentVarSetter.buildDisplayNameVar));
```
Comment (bsideup) [inline src/test/java/org/jenkinsci/plugins/buildnamesetter/BuildNameSetterTest.java]: Instead, you can test that environment variable was set. Presence of Action doesn't guarantee that environment variable will be set

File: src/main/java/org/jenkinsci/plugins/EnvironmentVarSetter.java [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+
+    public static void setVar(AbstractBuild build, String key, String value, PrintStream logger) {
+        EnvironmentVarSetter action = build.getAction(EnvironmentVarSetter.class);
+        if (action == null) {
+            action = new EnvironmentVarSetter(key, value, logger);
+        }
+        else {
+            action.setVar(key, value);
+        }
+        build.addAction(action);
```
Comment (bsideup) [inline src/main/java/org/jenkinsci/plugins/EnvironmentVarSetter.java]: I'm not 100% sure, but AFAIK this will add **another** Action if action is already presented.

@oleg-nenashev can you verify it?

Comment (Le0Michine) [inline src/main/java/org/jenkinsci/plugins/EnvironmentVarSetter.java]: I've just checked that, you are right, I'll fix

File: src/main/java/org/jenkinsci/plugins/EnvironmentVarSetter.java [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+            log.println("Get var: " + key + "=" + envVars.get(key));
+            return envVars.get(key);
+        }
+        else {
+            log.println("Var '" + key + "' doesn't exist");
+            return "";
+        }
+    }
+
+    public void buildEnvVars(AbstractBuild<?, ?> abstractBuild, EnvVars envVars) {
```
Comment (bsideup) [inline src/main/java/org/jenkinsci/plugins/EnvironmentVarSetter.java]: put `@Override` on any overridden method

**Comment by Le0Michine** [comment]
@bsideup  @oleg-nenashev ping
